### PR TITLE
Add check version functionality

### PR DESCRIPTION
**Pull request by bnkai**
Adds initial support for new version checking. Mentioned also in #288 
As it is now it uses the github api to get the shorthash commits from the repo tags.
It matches the current version ( `v0.0.0-alpha` or 'v0.0.0-alpha_dev' ) to the respective repo tags and gets the related commit hash from there.If it's different than the local that means that a new release build is available. 
It first prints a message in the console when the app is started and you can see it also when you open the about tab from the settings.An extra button is also available for manual checking.

for dev testing the build , since version is added through TRAVIS you need to add it to the ldflags in the Makefile
`-X 'github.com/stashapp/stash/pkg/api.version=v0.0.0-alpha_dev' `
or
`-X 'github.com/stashapp/stash/pkg/api.version=v0.0.0-alpha'`

console

ui

### CODE REVIEW

**Comment by Leopere** [comment]
Eventually, I really would like to update automagically by default.  Potentially allow people to disable the update mechanism via some command line launch argument or something not in the UI.

**Comment by WithoutPants** [comment]
Reviewed but not yet tested.

File: pkg/api/check_version.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+	if err != nil {
+		return "", fmt.Errorf("Github API request failed: %s", err)
+	} else {
+
+		defer response.Body.Close()
+
+		data, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return "", fmt.Errorf("Github API read respone failed: %s", err)
```
Comment (WithoutPants) [inline pkg/api/check_version.go]: Typo - respon**s**e

File: pkg/api/check_version.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+//gets latest version (git commit hash) from github API
+//the repo's tags are used to find the latest version
+//of the "master" or "develop" branch
+func GetLatestVersion(shortHash bool) (string, error) {
+
+	branch, _, _ := GetVersion()
+	if branch == "" {
+		return "", fmt.Errorf("Stash doesn't have a version.Version check not supported.")
```
Comment (WithoutPants) [inline pkg/api/check_version.go]: Add space between sentences.

File: Makefile [hunk 3]
```diff
@@ -6,7 +6,7 @@ endif
 build:
 	$(eval DATE := $(shell go run scripts/getDate.go))
 	$(eval GITHASH := $(shell git rev-parse --short HEAD))
-	$(SET) CGO_ENABLED=1 $(SEPARATOR) go build -mod=vendor -v -ldflags "-X 'github.com/stashapp/stash/pkg/api.buildstamp=$(DATE)' -X 'github.com/stashapp/stash/pkg/api.githash=$(GITHASH)'"
+	$(SET) CGO_ENABLED=1 $(SEPARATOR) go build -mod=vendor -v -ldflags "-X 'github.com/stashapp/stash/pkg/api.buildstamp=$(DATE)' -X 'github.com/stashapp/stash/pkg/api.githash=$(GITHASH)' "
```
Comment (WithoutPants) [inline Makefile]: Why is this change necessary?

**Comment by WithoutPants** [comment]
Since we are getting the URL anyway, is there any reason we can't or shouldn't include a direct link to the latest download in the about page?

File: pkg/api/check_version.go (hunk 1 above)
Comment (bnkai) [reply pkg/api/check_version.go]: fiixe

File: pkg/api/check_version.go (hunk 2 above)
Comment (bnkai) [reply pkg/api/check_version.go]: fixed

File: Makefile (hunk 3 above)
Comment (bnkai) [reply Makefile]: Sorry  , missed that. It was leftover from manual editing the Make file to add the version and so a space was left there.

**Comment by bnkai** [comment]
> Since we are getting the URL anyway, is there any reason we can't or shouldn't include a direct link to the latest download in the about page?

The API endpoint i am quering is this one [tags](https://api.github.com/repos/stashapp/stash/tags).
It only has the versions and the URL of the zipfile , tarball **source code** no releases.
There is also the [releases](https://api.github.com/repos/stashapp/stash/releases) one that can be queried ,  it is more complex to parse though.
I'll have a look at it , switching to WIP till i settle this.

**Comment by ueaslsef** [comment]
The download path should be rather straight forward: `https://github.com/stashapp/stash/releases/download/$version/stash-$platform`. Maybe we could cheat and skip the releases-api parsing all together and just use this until we found a better way? It could even be a static string inserted by the ci on build time.

**Comment by bnkai** [comment]
@ueaslsef  since we are doing it we might as well go all the way  ..
The releases API has extra stuff also like times downloaded or date released. Do we want any of that ?

**Comment by bnkai** [comment]
squashed and rebased , also fixed a merge conflict due to latest commits.
this the latest version with the proper release added as download link.
releases

File: pkg/api/check_version.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	}
+	return latestVersion, latestRelease, nil
+
+}
+
+func printLatestVersion() {
+	_, githash, _ = GetVersion()
+	latest, _, err := GetLatestVersion(true)
+	if err != nil {
+		logger.Errorf("Couldn't find latest version :%s\n", err)
```
Comment (WithoutPants) [inline pkg/api/check_version.go]: Change to `version: %s`. Looks weird in the log currently. The `\n` is unnecessary and pollutes the output.

File: ui/v2/src/components/Settings/SettingsAboutPanel.tsx [hunk 5]
```diff
@@ -26,32 +28,72 @@ export const SettingsAboutPanel: FunctionComponent<IProps> = (props: IProps) =>
...
           <tr>
-            <td>Build hash:</td>
-            <td>{data.version.hash}</td>
+            < td>Latest Version Build Hash: </td>
+            < td>{maybeRenderLatestVersion()} </td>
           </tr>
           <tr>
-            <td>Build time:</td>
-            <td>{data.version.build_time}</td>
+            <td><Button minimal={true} onClick={() => refetch()} text="Check for new version!" /></td>
```
Comment (WithoutPants) [inline ui/v2/src/components/Settings/SettingsAboutPanel.tsx]: Having `minimal={true}` makes the button look weird in this context when not moused over. I'd get rid of it. I'm not sure that the exclamation mark is necessary, but that's just a subjective aesthetic opinion.

File: pkg/api/check_version.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	} else {
+		if githash == latest {
+			logger.Infof("Version: (%s) is already the latest released.\n", latest)
```
Comment (WithoutPants) [inline pkg/api/check_version.go]: Remove `\n`

File: pkg/api/check_version.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+		} else {
+			logger.Infof("New version: (%s) available.\n", latest)
```
Comment (WithoutPants) [inline pkg/api/check_version.go]: Remove `\n`.

File: ui/v2/src/components/Settings/SettingsAboutPanel.tsx [hunk 8]
```diff
@@ -26,32 +28,72 @@ export const SettingsAboutPanel: FunctionComponent<IProps> = (props: IProps) =>
...
     return (
-      <>
       <HTMLTable>
         <tbody>
-          {maybeRenderTag()}
           <tr>
-            <td>Build hash:</td>
-            <td>{data.version.hash}</td>
+            < td>Latest Version Build Hash: </td>
+            < td>{maybeRenderLatestVersion()} </td>
```
Comment (WithoutPants) [inline ui/v2/src/components/Settings/SettingsAboutPanel.tsx]: Spaces before the `td` tag need to be fixed.

File: ui/v2/src/components/Settings/SettingsAboutPanel.tsx [hunk 9]
```diff
@@ -26,32 +28,72 @@ export const SettingsAboutPanel: FunctionComponent<IProps> = (props: IProps) =>
...
+        </>
+      );
+    }
+
+    return (
+      <>{dt.latestversion.shorthash}</>
+    );
+  }
+
+  function RenderLatestVersion() {
```
Comment (WithoutPants) [inline ui/v2/src/components/Settings/SettingsAboutPanel.tsx]: First letter should be lowercase.

File: ui/v2/src/components/Settings/SettingsAboutPanel.tsx [hunk 10]
```diff
@@ -26,32 +28,72 @@ export const SettingsAboutPanel: FunctionComponent<IProps> = (props: IProps) =>
...
   return (
     <>
       <H4>About</H4>
       {!data || loading ? <Spinner size={Spinner.SIZE_LARGE} /> : undefined}
-      {!!error ? <span>error.message</span> : undefined}
+      {!!error ? <span>{error.message}</span> : undefined}
+      {!!er ? <span>{er.message}</span> : undefined}
       {renderVersion()}
+      {!dt || ld || networkStatus == 4 ? <Spinner size={Spinner.SIZE_SMALL} /> : undefined}
+      {RenderLatestVersion()}
```
Comment (WithoutPants) [inline ui/v2/src/components/Settings/SettingsAboutPanel.tsx]: I think these variables (`dt`, `ld`, `er`) need to be better named.

I would make it so that if the spinner is showing (ie the query is in progress) then `renderLatestVersion` is not called.

File: pkg/api/check_version.go (hunk 4 above)
Comment (bnkai) [reply pkg/api/check_version.go]: i think i added it for the console, removed anyway

File: ui/v2/src/components/Settings/SettingsAboutPanel.tsx (hunk 5 above)
Comment (bnkai) [reply ui/v2/src/components/Settings/SettingsAboutPanel.tsx]: i tried the combo minimal and ! and plain button without the !
i'll change back to plain button

File: ui/v2/src/components/Settings/SettingsAboutPanel.tsx (hunk 10 above)
Comment (bnkai) [reply ui/v2/src/components/Settings/SettingsAboutPanel.tsx]: should be covered with latest commit ....

File: pkg/api/check_version.go (hunk 4 above)
Comment (WithoutPants) [reply pkg/api/check_version.go]: I need to proofread my feedback. This was supposed to be two separate items: the colon position and the `\n` character. Currently, the log looks like this:
```
time="2020-01-05T09:54:31+11:00" level=error msg="Couldn't find latest version :Stash doesn't have a version. Version check not supported."
```

Should be `version: %s` rather than `version :%s`.

Comment (bnkai) [reply pkg/api/check_version.go]: missed that, fixed.
